pip/pip-379.md

```diff
@@ -0,0 +1,115 @@
+
+The Key_Shared subscription doesn't prevent using any methods in the consumer API. For example, the application might call `negativeAcknowledge` or the `redeliverUnacknowledgedMessages` method. When messages are scheduled for delivery due to these methods, they will get redelivered as soon as possible. There's no ordering guarantee in these cases, however the guarantee of delivering a message key to a single consumer at a time will continue to be preserved.
+
+### Future work in needed for supporting key-based ordering with negative acknowledgements
+
+The updated contract explicitly states that it is not possible to retain key-based ordering of messages when negative acknowledgements are used. Changing this is out of scope for PIP-379. A potential future solution for handling this would be to modify the client so that when a message is negatively acknowledged, it would also reject all further messages with the same key until the original message gets redelivered. It's already possible to attempt to implement this in client-side code. However, a proper solution would require support on the broker side to block further delivery of the specific key when there are pending negatively acknowledged messages until all negatively acknowledged messages for that particular key have been acknowledged by the consumer. This solution is out of scope for PIP-379. A future implementation to address these problems could build upon PIP-379 concepts such as "draining hashes" and extend that to cover the negative acknowledgement scenarios.
+
+### High-Level implementation plan
+
+The proposed solution introduces a "draining hashes" concept to efficiently manage message ordering in Key_Shared subscriptions:
+
+1. When consumer hash ranges change (e.g., a consumer joins or leaves), affected hashes of pending messages are added to a "draining hashes" set.
+2. Following messages with hashes in the "draining hashes" set are blocked from further delivery until pending messages are processed.
+3. A reference counter tracks pending messages for each hash in the "draining hashes" set.
+4. As messages are acknowledged or consumers disconnect and therefore get removed from pending messages, the reference counter is decremented.
+5. When the reference counter reaches zero, the hash is removed from the set, allowing new message delivery.
+6. Consumer hash assignments may change multiple times, and a draining hash might be reassigned to the original consumer. The draining hash data structure contains information about the draining consumer. When a message is attempted for delivery, the system can check if the target consumer is the same as the draining consumer. If they match, there's no need to block the hash. The implementation should also remove such hashes from the draining hashes set. This "lazy" approach reduces the need for actively scanning all draining hashes whenever hash assignments change.
```

Comment:
When a consumer disconnects, the messages return to the redelivery queue. There are currently some race conditions related to the timing of when these messages are delivered. This was discussed in https://github.com/apache/pulsar/issues/23307#issuecomment-2360183793 . 

In the current version of PIP-379, the "Updated contract of preserving ordering" is simply "In Key_Shared subscriptions, messages with the same key are delivered and allowed to be in unacknowledged state to only one consumer at a time.". In the implementation of PIP-379, this is an [invariant](https://en.wikipedia.org/wiki/Invariant_(mathematics)#Invariants_in_computer_science) which must hold at all times. Since the invariant is simple, there's much better chances to succeed in holding this invariant at all times in the system. I believe that taking this approach will help detect and resolve possible race conditions that would break this invariant.

Reply:
Okay. I think I understand your point.
But my question is which consumer can receive redelivery messages that were held by the disconnected consumer?

However, I began to think that it would simply be the consumer that the selector returns.

Reply:
> Okay. I think I understand your point. But my question is which consumer can receive redelivery messages that were held by the disconnected consumer?
> 
> However, I began to think that it would simply be the consumer that the selector returns.

@equanz Yes, messages are always delivered to the consumer that the selector returns.